### PR TITLE
[Refactor] Execute 'pick rowsets' before applying for permits for a compaction task

### LINKED ISSUE #4882: [Proposal] Execute 'pick rowsets' before applying for permits for a compaction task
The current compaction mechanism is that there is a producer thread that has been producing compaction tasks, and the selected tablet must apply for `permits`. When a tablet could hold `permits`, compaction task for this tablet will be submitted to  thread pool. We take compaction score as `permits` which is used for limiting memory consumption. However,  `pick_rowset_to_compaction()` will be executed before the file merge in compaction thread, and the number of segment files that actually perform the merge operation is smaller than compaction score. In addition, it is also possible that compaction task exits directly because the tablet doesn't meet the requirements of compaction. 

Can we execute 'pick rowsets' before applying for permits for a compaction task, calculate the number of segment files that actually participate in the merge operation, and take this number as `permits`.

Based on test results, we found that`pick_rowset_to_compaction()` execution time for a tablet is on the time scale of milliseconds. Therefore, there is little pressure on the producer thread.

```
1112 16:00:08.686122 (+   982us) cumulative_compaction.cpp:50] rowsets picked
1112 16:00:08.355522 (+  1033us) cumulative_compaction.cpp:50] rowsets picked
1112 16:00:08.355372 (+   930us) cumulative_compaction.cpp:50] rowsets picked
1112 16:00:09.707385 (+   990us) cumulative_compaction.cpp:50] rowsets picked
1112 16:00:32.218772 (+  1305us) cumulative_compaction.cpp:50] rowsets picked
1112 16:00:32.212823 (+  1212us) cumulative_compaction.cpp:50] rowsets picked
1112 16:00:32.077322 (+  2411us) cumulative_compaction.cpp:50] rowsets picked
1112 16:00:33.534394 (+  2484us) cumulative_compaction.cpp:50] rowsets picked
1112 16:00:52.350023 (+  1181us) cumulative_compaction.cpp:50] rowsets picked
1112 16:00:54.011435 (+  1212us) cumulative_compaction.cpp:50] rowsets picked
1112 16:01:14.845957 (+   294us) cumulative_compaction.cpp:50] rowsets picked
1112 16:00:17.865887 (+    97us) base_compaction.cpp:44] rowsets picked
1112 16:01:35.193204 (+   206us) cumulative_compaction.cpp:50] rowsets picked
1112 16:01:39.244388 (+   198us) cumulative_compaction.cpp:50] rowsets picked
1112 16:02:18.931984 (+  1627us) cumulative_compaction.cpp:50] rowsets picked
1112 16:02:19.014592 (+  1394us) cumulative_compaction.cpp:50] rowsets picked
1112 16:03:03.405753 (+   160us) cumulative_compaction.cpp:50] rowsets picked
1112 16:03:04.777374 (+   350us) cumulative_compaction.cpp:50] rowsets picked
1112 16:03:07.016017 (+   164us) cumulative_compaction.cpp:50] rowsets picked
1112 15:59:12.573256 (+   205us) base_compaction.cpp:44] rowsets picked
1112 16:02:18.932822 (+  2213us) cumulative_compaction.cpp:50] rowsets picked
1112 16:03:08.975090 (+   339us) cumulative_compaction.cpp:50] rowsets picked
1112 16:02:18.932387 (+  1885us) cumulative_compaction.cpp:50] rowsets picked
1112 16:03:11.285206 (+   402us) cumulative_compaction.cpp:50] rowsets picked
1112 16:03:14.799639 (+   119us) cumulative_compaction.cpp:50] rowsets picked
1112 16:03:20.540975 (+   148us) cumulative_compaction.cpp:50] rowsets picked
1112 16:03:23.908919 (+   198us) cumulative_compaction.cpp:50] rowsets picked
1112 16:03:32.000026 (+   133us) cumulative_compaction.cpp:50] rowsets picked
1112 16:03:13.347364 (+  2796us) cumulative_compaction.cpp:50] rowsets picked
1112 16:03:13.722429 (+  2046us) cumulative_compaction.cpp:50] rowsets picked
1112 16:03:43.050516 (+   410us) cumulative_compaction.cpp:50] rowsets picked
1112 16:03:43.849807 (+   387us) cumulative_compaction.cpp:50] rowsets picked
1112 16:02:17.914220 (+   469us) base_compaction.cpp:44] rowsets picked
1112 16:03:48.663824 (+   336us) cumulative_compaction.cpp:50] rowsets picked
1112 16:03:49.778246 (+   241us) cumulative_compaction.cpp:50] rowsets picked
1112 16:03:52.141355 (+   468us) cumulative_compaction.cpp:50] rowsets picked
1112 16:04:20.670042 (+  1564us) cumulative_compaction.cpp:50] rowsets picked
1112 16:04:38.078296 (+   264us) cumulative_compaction.cpp:50] rowsets picked
1112 16:04:36.673882 (+  1177us) cumulative_compaction.cpp:50] rowsets picked
1112 16:04:37.060312 (+  1139us) cumulative_compaction.cpp:50] rowsets picked
1112 16:04:55.154987 (+  1215us) cumulative_compaction.cpp:50] rowsets picked
1112 16:04:55.403399 (+  1071us) cumulative_compaction.cpp:50] rowsets picked
1112 16:05:02.536426 (+  1551us) cumulative_compaction.cpp:50] rowsets picked
1112 16:05:32.090439 (+  1024us) cumulative_compaction.cpp:50] rowsets picked
1112 16:05:18.661826 (+  1306us) cumulative_compaction.cpp:50] rowsets picked
1112 16:05:18.990078 (+  1241us) cumulative_compaction.cpp:50] rowsets picked
1112 16:05:42.798424 (+   143us) cumulative_compaction.cpp:50] rowsets picked
1112 16:05:46.082527 (+   176us) cumulative_compaction.cpp:50] rowsets picked
1112 16:05:41.727956 (+  1584us) cumulative_compaction.cpp:50] rowsets picked
1112 16:06:01.083588 (+  1123us) cumulative_compaction.cpp:50] rowsets picked
1112 16:06:26.622230 (+   632us) cumulative_compaction.cpp:50] rowsets picked
1112 16:06:35.897463 (+    92us) cumulative_compaction.cpp:50] rowsets picked
```

### PR DESCRIPTION
**Pull request by weizuo93**
## Proposed changes

The current compaction mechanism is that there is a producer thread that has been producing compaction tasks, and the selected tablet must apply for `permits`. When a tablet could hold `permits`, compaction task for this tablet will be submitted to  thread pool. We take compaction score as `permits` which is used for limiting memory consumption. However,  `pick_rowset_to_compaction()` will be executed before the file merge in compaction thread, and the number of segment files that actually perform the merge operation is smaller than compaction score. In addition, it is also possible that compaction task exits directly because the tablet doesn't meet the requirements of compaction. 

This patch optimizes and refactors the code of compaction, so that we can execute 'pick rowsets' before applying for permits for a compaction task, calculate the number of segment files that actually participate in the merge operation, and take this number as `permits`.

## Types of changes

What types of changes does your code introduce to Doris?
_Put an `x` in the boxes that apply_

- [] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [] Documentation Update (if none of the other choices apply)
- [x] Code refactor (Modify the code structure, format the code, etc...)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] I have create an issue on (Fix #4882 ), and have described the bug/feature there in detail
- [] Compiling and unit tests pass locally with my changes
- [] I have added tests that prove my fix is effective or that my feature works
- [x] If this change need a document change, I have updated the document
- [x] Any dependent changes have been merged

### CODE REVIEW

File: be/src/olap/base_compaction.cpp [hunk 1]
```diff
@@ -43,9 +50,26 @@ OLAPStatus BaseCompaction::compact() {
...
+OLAPStatus BaseCompaction::execute_compact() {
+    MutexLock lock(_tablet->get_base_lock(), TRY_LOCK);
+    if (!lock.own_lock()) {
+        LOG(WARNING) << "another base compaction is running. tablet=" << _tablet->full_name();
+        return OLAP_ERR_BE_TRY_BE_LOCK_ERROR;
+    }
+    TRACE("got base compaction lock");
+
+    if (_tablet->get_clone_occurred()) {
+        _tablet->set_clone_occurred(false);
```
Comment (morningman) [inline be/src/olap/base_compaction.cpp:67]: Add comment to explain why setting clone occurred here.

File: be/src/olap/base_compaction.cpp [hunk 2]
```diff
@@ -28,6 +28,13 @@ BaseCompaction::BaseCompaction(TabletSharedPtr tablet, const std::string& label,
 BaseCompaction::~BaseCompaction() { }
 
 OLAPStatus BaseCompaction::compact() {
+    RETURN_NOT_OK(prepare_compact());
```
Comment (morningman) [inline be/src/olap/base_compaction.cpp:31]: Who will call `BaseCompaction::compact()` now?

Comment (weizuo93) [reply be/src/olap/base_compaction.cpp:31]: > Who will call `BaseCompaction::compact()` now?

`BaseCompaction::compact()` will still be called in "be/src/http/action/compaction_action.cpp" for `/api/compaction/run`.

File: be/src/olap/base_compaction.cpp (hunk 1 above)
Comment (weizuo93) [reply be/src/olap/base_compaction.cpp:67]: OK.

**Review by morningman — APPROVED** [review]
LGTM

**Pull request merged**